lib/SILOptimizer/Transforms/AllocBoxToStack.cpp

```diff
@@ -802,67 +892,83 @@ specializePartialApply(SILOptFunctionBuilder &FuncBuilder,
+
+    if (Apply.getKind() == ApplySiteKind::PartialApplyInst) {
+      if (PAFrontier.empty()) {
+        ValueLifetimeAnalysis VLA(cast<PartialApplyInst>(Apply));
+        pass.CFGChanged |= !VLA.computeFrontier(
+            PAFrontier, ValueLifetimeAnalysis::AllowToModifyCFG);
+        assert(!PAFrontier.empty() &&
+               "partial_apply must have at least one use "
+               "to release the returned function");
+      }
 
-    // Insert destroys of the box at each point where the partial_apply becomes
-    // dead.
-    for (SILInstruction *FrontierInst : PAFrontier) {
-      SILBuilderWithScope Builder(FrontierInst);
-      Builder.createDestroyValue(PartialApply->getLoc(), Box);
+      // If this argument is promoted, it is a box that we're turning into an
+      // address because we've proven we can keep this value on the stack. The
+      // partial_apply had ownership of this box so we must now release it
+      // explicitly when the partial_apply is released.
+      // Insert destroys of the box at each point where the partial_apply
+      // becomes dead.
+      for (SILInstruction *FrontierInst : PAFrontier) {
+        SILBuilderWithScope Builder(FrontierInst);
+        Builder.createDestroyValue(Apply.getLoc(), Box);
+      }
     }
   }
 
-  SILBuilderWithScope Builder(PartialApply);
+  SILBuilderWithScope Builder(Apply.getInstruction());
 
-  // Build the function_ref and partial_apply.
-  SILValue FunctionRef = Builder.createFunctionRef(PartialApply->getLoc(),
+  // Build the function_ref and partial_apply/apply.
+  SILValue FunctionRef = Builder.createFunctionRef(Apply.getLoc(),
                                                    ClonedFn);
-  return Builder.createPartialApply(
-      PartialApply->getLoc(), FunctionRef, PartialApply->getSubstitutionMap(),
-      Args,
-      PartialApply->getType().getAs<SILFunctionType>()->getCalleeConvention());
+  if (Apply.getKind() == ApplySiteKind::PartialApplyInst) {
+    return Builder.createPartialApply(
+        Apply.getLoc(), FunctionRef, Apply.getSubstitutionMap(), Args,
+        cast<PartialApplyInst>(Apply.getInstruction())->getType()
+            .getAs<SILFunctionType>()->getCalleeConvention());
+  }
+  return Builder.createApply(
+        Apply.getLoc(), FunctionRef, Apply.getSubstitutionMap(), Args);
 }
 
-static void rewritePartialApplies(AllocBoxToStackState &pass) {
-  llvm::DenseMap<PartialApplyInst *, ArgIndexList> IndexMap;
+static void rewriteApplies(AllocBoxToStackState &pass) {
+  llvm::DenseMap<ApplySite, ArgIndexList> IndexMap;
   ArgIndexList Indices;
 
   // Build a map from partial_apply to the indices of the operands
   // that will be promoted in our rewritten version.
   for (auto *O : pass.PromotedOperands) {
-    auto CalleeArgIndexNumber = ApplySite(O->getUser()).getCalleeArgIndex(*O);
+    auto User = O->getUser();
+    auto Apply = ApplySite(User);
+    assert(isa<PartialApplyInst>(User) || isa<ApplyInst>(User));
```

Comment:
My thought: use an exhaustive switch below to make people explicitly crash on unhandled types. Will make sure people update this.